File: go/modifying_sequence.go
package main

import "math/rand"

func CopyIf(dest []int, src []int, f func(int) bool) int {
	i := 0
	endOfDest := len(dest)
	for _, e := range src {
		if f(e) {
			dest[i] = e
			i++
			if i == endOfDest {
				break
			}
		}
	}
	return i
}

func CopyN(dest []int, src []int, count int) []int {
	for i := 0; i < count; i++ {
		dest[i] = src[i]
	}
	return dest
}

func CopyBackward(dest []int, src []int) int {
	i, j := len(dest)-1, len(src)-1
	for ; i >= 0 && j >= 0; i-- {
		dest[i] = src[j]
		j--
	}
	if i < j { // can't copy all elements from src to dest
		return len(dest) // so we have copied len(dest) elements
	} else {
		return len(src)
	}
}

func Fill(dest []int, value int) {
	for i := range dest {
		dest[i] = value
	}
}

func FillN(dest []int, count int, value int) {
	var end int
	if count < len(dest) {
		end = count
	} else {
		end = len(dest)
	}
	for i := 0; i < end; i++ {
		dest[i] = value
	}
}

func Transform(dest []int, src []int, f func(int) int) []int {
	var minLength int
	if len(dest) < len(src) {
		minLength = len(dest)
	} else {
		minLength = len(src)
	}
	for i := 0; i < minLength; i++ {
		dest[i] = f(src[i])
	}
	return dest
}

func Generate(dest []int, g func() int) {
	for i := range dest {
		dest[i] = g()
	}
}

func GenerateN(dest []int, count int, g func() int) {
	for i := 0; i < count; i++ {
		dest[i] = g()
	}
}

func Remove(src []int, match int) []int {
	var dest []int
	for _, e := range src {
		if e != match {
			dest = append(dest, e)
		}
	}
	return dest
}

func RemoveIf(src []int, f func(int) bool) []int {
	var dest []int
	for _, e := range src {
		if !f(e) {
			dest = append(dest, e)
		}
	}
	return dest
}

func RemoveCopy(dest []int, src []int, match int) []int {
	var tmp []int
	for _, e := range src {
		if e != match {
			tmp = append(tmp, e)
		}
	}
	copy(dest, tmp)
	return dest
}

func RemoveCopyIf(dest []int, src []int, f func(int) bool) []int {
	var tmp []int
	for _, e := range src {
		if !f(e) {
			tmp = append(tmp, e)
		}
	}
	copy(dest, tmp)
	return dest
}

func Replace(src []int, before int, after int) {
	for i := range src {
		if src[i] == before {
			src[i] = after
		}
	}
}

func ReplaceIf(src []int, f func(int) bool, after int) {
	for i, e := range src {
		if f(e) {
			src[i] = after
		}
	}
}

func ReplaceCopy(dest []int, src []int, before int, after int) []int {
	var tmp []int
	for _, e := range src {
		if e != before {
			tmp = append(tmp, e)
		} else {
			tmp = append(tmp, after)
		}
	}
	copy(dest, tmp)
	return dest
}

func ReplaceCopyIf(dest []int, src []int, f func(int) bool, after int) []int {
	var tmp []int
	for _, e := range src {
		if !f(e) {
			tmp = append(tmp, e)
		} else {
			tmp = append(tmp, after)
		}
	}
	copy(dest, tmp)
	return dest
}

func Swap(a *int, b *int) {
	*a, *b = *b, *a
}

func SwapRanges(a []int, b []int, size int) {
	if size > len(a) || size > len(b) {
		if len(a) > len(b) {
			size = len(b)
		} else {
			size = len(a)
		}
	}
	for i := 0; i < size; i++ {
		a[i], b[i] = b[i], a[i]
	}
}

func Reverse(src []int) {
	size := len(src) - 1
	for i := 0; i < size/2+1; i++ {
		src[i], src[size-i] = src[size-i], src[i]
	}
}

func ReverseCopy(dest []int, src []int) []int {
	var size int
	lenOfSrc := len(src)
	if lenOfSrc > len(dest) {
		size = len(dest)
	} else {
		size = lenOfSrc
	}
	lenOfSrc--
	for i := 0; i < size; i++ {
		dest[i] = src[lenOfSrc-i]
	}
	return dest
}

func Rotate(dest []int, pivot int) {
	end := len(dest)
	if pivot <= 0 || pivot >= end || end <= 0 {
		return
	}
	newStart, newPlace := pivot, pivot
	start := 0
	for {
		dest[start], dest[newPlace] = dest[newPlace], dest[start]
		start++
		newPlace++
		if newPlace == end {
			newPlace = newStart
		} else if start == newStart {
			newStart = newPlace
		}
		if start == newPlace {
			break
		}
	}
}

func RotateCopy(dest []int, src []int, pivot int) []int {
	n := copy(dest, src[pivot:])
	copy(dest[n:], src[:pivot])
	return dest
}

func Shuffle(dest []int) {
	end := len(dest)
	for i := range dest {
		rd := rand.Intn(end)
		dest[i], dest[rd] = dest[rd], dest[i]
	}
}

func Unique(dest []int) []int {
	end := len(dest)
	head := 0
	for i := 1; i < end; i++ {
		if dest[head] != dest[i] {
			head++
			dest[head] = dest[i]
		}
	}
	return dest[:head+1]
}

func UniqueCopy(src []int) []int {
	end := len(src)
	dest := make([]int, 0)
	if end <= 0 {
		return dest
	}
	head := 0
	dest = append(dest, src[head])
	for i := 1; i < end; i++ {
		if src[head] != src[i] {
			head = i
			dest = append(dest, src[i])
		}
	}
	return dest
}

File: ruby/modifying_sequence.rb
# Implementations of modifying sequence algorithms
module ModifyingSequence

def copy(iterable, start, stop)
  iterable.drop(start).take(stop - start)
end

def copy_if(iterable, &block)
  iterable.select(&block)
end

def copy_n(iterable, count)
  iterable.take(count)
end

def copy_backward(iterable, start, stop)
  copy(iterable, start, stop).reverse_each.to_a
end

alias_method :move, :copy
alias_method :move_backward, :copy_backward

def fill(seq, value)
  seq.each_index{|i| seq[i] = value}
end

def fill_n(seq, value, count)
  min = (count < seq.size ? count : seq.size)
  (0...min).each{|i| seq[i] = value}
end

def transform(iterable, &block)
  iterable.map(&block)
end

def generate(seq)
  seq.each_index{|i| seq[i] = yield }
end

def generate_n(seq, count)
  min = (count < seq.size ? count : seq.size)
  (0...min).each{|i| seq[i] = yield}
  seq
end

def remove(iterable, match)
  iterable.reject{|x| x == match}
end

def remove_if(iterable, &block)
  iterable.reject(&block)
end

# implement remove_copy as remove
alias_method :remove_copy, :remove

# implement remove_copy_if as remove_if
alias_method :remove_copy_if, :remove_if

def replace(iterable, before, after)
  iterable.map{|x| x == before ? after : x}
end

def replace_if(iterable, after)
  iterable.map{|x| (yield x) ? after : x}
end

# implement replace_copy as replace
alias_method :replace_copy, :replace

# implement replace_copy_if as replace_if
alias_method :replace_copy_if, :replace_if

def swap(a, b)
  return b, a
end

# implement iter_swap as swap
alias_method :iter_swap, :swap

def swap_range(a, b, start, stop)
  max_size = (a.size > b.size ? a.size : b.size)
  return if start < 0 || start >= stop || stop > max_size
  (start...stop).each{|i| a[i], b[i] = b[i], a[i]}
end

def reverse_in_place(seq, start, stop)
  (start..((start+stop)/2).to_i).each{
    |i| seq[i], seq[start+stop-i] = seq[start+stop-i], seq[i]
  }
end

def reverse(seq)
  stop = seq.size - 1
  reverse_in_place(seq, 0, stop)
end

# implement reverse_copy as reverse
alias_method :reverse_copy, :reverse

def rotate(seq, pivot)
  reverse_in_place(seq, 0, pivot-1)
  reverse_in_place(seq, pivot, seq.length-1)
  reverse_in_place(seq, 0, seq.length-1)
  seq
end

def rotate_copy(seq, pivot)
  seq[pivot...seq.length] + seq[0...pivot]
end

def shuffle(seq)
  srand Time.now.to_i
  (0...seq.length).each do |i|
    random = rand(i)
    seq[i], seq[random] = seq[random], seq[i]
  end
end

def unique(seq)
  return if seq.length <= 1
  current = seq[0]
  seq[1...seq.length].reject do |x|
    same = (x == current)
    current = x unless same
    same
  end.unshift(seq[0])
end

# implement unique_copy as unique
alias_method :unique_copy, :unique

end
